src/modules/sync/utils/htmlRenderers.ts

```diff
@@ -0,0 +1,638 @@
+        }</p>`;
+      }
+      return getValueOrIdentifier(value);
+    default:
+      return null;
+  }
+};
+
+const renderDeletedEntity = (entityCodename: string) =>
+  `<div class="entity-detail"><div class="entity-name removed">${entityCodename}</div></div>`;
+
+const renderAddedTypesOrSnippetsSectionData = <T extends RequiredCodename<TypeOrSnippet>>(
+  diff: Pick<DiffObject<RequiredCodename<T>>, "added">,
+) => diff.added.map(a => renderAddedTypeOrSnippetData(a, a.elements)).join("\n");
+
+const renderAddedTaxonomiesSectionData = <T extends TaxonomyModels.IAddTaxonomyRequestModel>(
+  diff: Pick<DiffObject<RequiredCodename<T>>, "added">,
+) => diff.added.map(a => renderAddedTaxonomyData(a, a.terms)).join("\n");
+
+const renderDeletedEntitiesSectionData = <T extends { codename: string }>(
+  diff: Pick<DiffObject<RequiredCodename<T>>, "deleted">,
+) =>
+  `${
+    [...diff.deleted]
+      .map(renderDeletedEntity)
+      .join("\n")
+  }`;
+
+const renderUpdatedEntitiesSectionData = <T extends { codename: string }>(
+  diff: Pick<DiffObject<RequiredCodename<T>>, "updated">,
+) =>
+  [...diff.updated]
+    .filter(([, p]) => p.length > 0)
+    .flatMap(renderUpdatedEntity)
+    .join("\n");
+
+const renderUpdatedEntity = (
+  [entityCodename, patchOps]: Readonly<[string, Readonly<PatchOperation[]>]>,
+) =>
+  `<div class="entity-detail"><div class="entity-name" onClick="toggleVisibility('${entityCodename}')">${entityCodename}</div><div class="entity-operations" style="display: none" id="${entityCodename}">${
+    patchOps
+      .flatMap(renderPatchOp)
+      .join("\n")
+  }</div></div>`;
+
+const renderEntitySection = (
+  entityType: EntityType,
+  changes: DiffObject<unknown>,
+) => {
+  const { added, updated, deleted } = changes;
+  const entityTypeNameMap: ReadonlyMap<typeof entityType, string> = new Map([
+    ["snippets", "Snippets"],
+    ["taxonomies", "Taxonomy groups"],
+    ["types", "Content types"],
+  ]);
+
+  return `
+        <div class="entity-section">
+          <div class="entity-section-header" onclick="toggleVisibility('${entityType}')">
+              <div>${entityTypeNameMap.get(entityType)}</div>
+              <div class="num-modified push">✎ {{num_${entityType}_updated}}</div>
+              <div class="num-added">+ {{num_${entityType}_added}}</div>
+              <div class="num-removed">− {{num_${entityType}_removed}}</div>
+          </div>
+          <div id="${entityType}" class="entity-section-content">
+              ${
+    [...updated.values()].filter(v => v.length).length
+      ? renderUpdatedEntitySection(entityType)
+      : ""
+  }
+              <div class="added-and-deleted">
+                  ${deleted.size ? renderRemovedEntitySection(entityType) : ""}
+                  ${added.length ? renderAddedEntitySection(entityType) : ""}
+              </div>
+          </div>
+        </div>
+      `;
+};
+
+const getCombinedOpLength = (ops: DiffObject<unknown>) =>
+  [
```

Comment:
how about this? You wouldn't have to copy objects creating new arrays, so you can save some complexity :)
```ts
const getCombinedOpLength = (ops: DiffObject<unknown>) =>
  ops.added.length + ops.deleted.size + [...ops.updated.values()].filter(v => v.length).length;
```

Reply:
makes sense, fixed